// mongo-errors/index.js
exports.class = {};

exports.OK = 0;
exports.InternalError = 1;
exports.BadValue = 2;
exports.OBSOLETE_DuplicateKey = 3;
exports.NoSuchKey = 4;
exports.GraphContainsCycle = 5;
exports.HostUnreachable = 6;
exports.HostNotFound = 7;
exports.UnknownError = 8;
exports.FailedToParse = 9;
exports.CannotMutateObject = 10;
exports.UserNotFound = 11;
exports.UnsupportedFormat = 12;
exports.Unauthorized = 13;
exports.TypeMismatch = 14;
exports.Overflow = 15;
exports.InvalidLength = 16;
exports.ProtocolError = 17;
exports.AuthenticationFailed = 18;
exports.CannotReuseObject = 19;
exports.IllegalOperation = 20;
exports.EmptyArrayOperation = 21;
exports.InvalidBSON = 22;
exports.AlreadyInitialized = 23;
exports.LockTimeout = 24;
exports.RemoteValidationError = 25;
exports.NamespaceNotFound = 26;
exports.IndexNotFound = 27;
exports.PathNotViable = 28;
exports.NonExistentPath = 29;
exports.InvalidPath = 30;
exports.RoleNotFound = 31;
exports.RolesNotRelated = 32;
exports.PrivilegeNotFound = 33;
exports.CannotBackfillArray = 34;
exports.UserModificationFailed = 35;
exports.RemoteChangeDetected = 36;
exports.FileRenameFailed = 37;
exports.FileNotOpen = 38;
exports.FileStreamFailed = 39;
exports.ConflictingUpdateOperators = 40;
exports.FileAlreadyOpen = 41;
exports.LogWriteFailed = 42;
exports.CursorNotFound = 43;
exports.UserDataInconsistent = 45;
exports.LockBusy = 46;
exports.NoMatchingDocument = 47;
exports.NamespaceExists = 48;
exports.InvalidRoleModification = 49;
exports.ExceededTimeLimit = 50;
exports.ManualInterventionRequired = 51;
exports.DollarPrefixedFieldName = 52;
exports.InvalidIdField = 53;
exports.NotSingleValueField = 54;
exports.InvalidDBRef = 55;
exports.EmptyFieldName = 56;
exports.DottedFieldName = 57;
exports.RoleModificationFailed = 58;
exports.CommandNotFound = 59;
exports.OBSOLETE_DatabaseNotFound = 60;
exports.ShardKeyNotFound = 61;
exports.OplogOperationUnsupported = 62;
exports.StaleShardVersion = 63;
exports.WriteConcernFailed = 64;
exports.MultipleErrorsOccurred = 65;
exports.ImmutableField = 66;
exports.CannotCreateIndex = 67;
exports.IndexAlreadyExists = 68;
exports.AuthSchemaIncompatible = 69;
exports.ShardNotFound = 70;
exports.ReplicaSetNotFound = 71;
exports.InvalidOptions = 72;
exports.InvalidNamespace = 73;
exports.NodeNotFound = 74;
exports.WriteConcernLegacyOK = 75;
exports.NoReplicationEnabled = 76;
exports.OperationIncomplete = 77;
exports.CommandResultSchemaViolation = 78;
exports.UnknownReplWriteConcern = 79;
exports.RoleDataInconsistent = 80;
exports.NoMatchParseContext = 81;
exports.NoProgressMade = 82;
exports.RemoteResultsUnavailable = 83;
exports.DuplicateKeyValue = 84;
exports.IndexOptionsConflict = 85;
exports.IndexKeySpecsConflict = 86;
exports.CannotSplit = 87;
exports.SplitFailed_OBSOLETE = 88;
exports.NetworkTimeout = 89;
exports.CallbackCanceled = 90;
exports.ShutdownInProgress = 91;
exports.SecondaryAheadOfPrimary = 92;
exports.InvalidReplicaSetConfig = 93;
exports.NotYetInitialized = 94;
exports.NotSecondary = 95;
exports.OperationFailed = 96;
exports.NoProjectionFound = 97;
exports.DBPathInUse = 98;
exports.CannotSatisfyWriteConcern = 100;
exports.OutdatedClient = 101;
exports.IncompatibleAuditMetadata = 102;
exports.NewReplicaSetConfigurationIncompatible = 103;
exports.NodeNotElectable = 104;
exports.IncompatibleShardingMetadata = 105;
exports.DistributedClockSkewed = 106;
exports.LockFailed = 107;
exports.InconsistentReplicaSetNames = 108;
exports.ConfigurationInProgress = 109;
exports.CannotInitializeNodeWithData = 110;
exports.NotExactValueField = 111;
exports.WriteConflict = 112;
exports.InitialSyncFailure = 113;
exports.InitialSyncOplogSourceMissing = 114;
exports.CommandNotSupported = 115;
exports.DocTooLargeForCapped = 116;
exports.ConflictingOperationInProgress = 117;
exports.NamespaceNotSharded = 118;
exports.InvalidSyncSource = 119;
exports.OplogStartMissing = 120;
exports.DocumentValidationFailure = 121; // Only for the document validator on collections.
exports.OBSOLETE_ReadAfterOptimeTimeout = 122;
exports.NotAReplicaSet = 123;
exports.IncompatibleElectionProtocol = 124;
exports.CommandFailed = 125;
exports.RPCProtocolNegotiationFailed = 126;
exports.UnrecoverableRollbackError = 127;
exports.LockNotFound = 128;
exports.LockStateChangeFailed = 129;
exports.SymbolNotFound = 130;
exports.RLPInitializationFailed = 131;
exports.ConfigServersInconsistent = 132;
exports.FailedToSatisfyReadPreference = 133;
exports.ReadConcernMajorityNotAvailableYet = 134;
exports.StaleTerm = 135;
exports.CappedPositionLost = 136;
exports.IncompatibleShardingConfigVersion = 137;
exports.RemoteOplogStale = 138;
exports.JSInterpreterFailure = 139;
exports.InvalidSSLConfiguration = 140;
exports.SSLHandshakeFailed = 141;
exports.JSUncatchableError = 142;
exports.CursorInUse = 143;
exports.IncompatibleCatalogManager = 144;
exports.PooledConnectionsDropped = 145;
exports.ExceededMemoryLimit = 146;
exports.ZLibError = 147;
exports.ReadConcernMajorityNotEnabled = 148;
exports.NoConfigMaster = 149;
exports.StaleEpoch = 150;
exports.OperationCannotBeBatched = 151;
exports.OplogOutOfOrder = 152;
exports.ChunkTooBig = 153;
exports.InconsistentShardIdentity = 154;
exports.CannotApplyOplogWhilePrimary = 155;
exports.NeedsDocumentMove = 156;
exports.CanRepairToDowngrade = 157;
exports.MustUpgrade = 158;
exports.DurationOverflow = 159;
exports.MaxStalenessOutOfRange = 160;
exports.IncompatibleCollationVersion = 161;

// Non-sequential error codes (for compatibility only)
exports.SocketException = 9001;
exports.RecvStaleConfig = 9996;
exports.NotMaster = 10107;
exports.CannotGrowDocumentInCappedNamespace = 10003;
exports.DuplicateKey = 11000;
exports.InterruptedAtShutdown = 11600;
exports.Interrupted = 11601;
exports.InterruptedDueToReplStateChange = 11602;
exports.OutOfDiskSpace = 14031;
exports.KeyTooLong = 17280;
exports.BackgroundOperationInProgressForDatabase = 12586;
exports.BackgroundOperationInProgressForNamespace = 12587;
exports.NotMasterOrSecondary = 13436;
exports.NotMasterNoSlaveOk = 13435;
exports.ShardKeyTooBig = 13334;
exports.SendStaleConfig = 13388;
exports.DatabaseDifferCase = 13297;
exports.OBSOLETE_PrepareConfigsFailed = 13104;

// SERVER-21428 explains the extra changes needed if error_class components change.
exports.class.NetworkError = [exports.HostUnreachable, exports.HostNotFound, exports.NetworkTimeout];
exports.class.Interruption = [exports.Interrupted,
                             exports.InterruptedAtShutdown,
                             exports.InterruptedDueToReplStateChange,
                             exports.ExceededTimeLimit];
exports.class.NotMasterError = [exports.NotMaster, exports.NotMasterNoSlaveOk];
exports.class.StaleShardingError =
            [exports.RecvStaleConfig, exports.SendStaleConfig, exports.StaleShardVersion, exports.StaleEpoch];
exports.class.WriteConcernError = [exports.WriteConcernFailed,
                                  exports.WriteConcernLegacyOK,
                                  exports.UnknownReplWriteConcern,
                                  exports.CannotSatisfyWriteConcern];
